lightning/src/ln/functional_tests.rs

```diff
@@ -7411,7 +7407,7 @@ fn test_data_loss_protect() {
 	tx_broadcaster = test_utils::TestBroadcaster{txn_broadcasted: Mutex::new(Vec::new())};
 	fee_estimator = test_utils::TestFeeEstimator { sat_per_kw: 253 };
 	persister = test_utils::TestPersister::new();
-	monitor = test_utils::TestChainMonitor::new(Some(&chain_source), &tx_broadcaster, &logger, &fee_estimator, &persister);
+	monitor = test_utils::TestChainMonitor::new(Some(&chain_source), &tx_broadcaster, &logger, &fee_estimator, &persister, &keys_manager);
```

Comment:
Hmm, at least here, maybe elsewhere - dont we create the TestKeysInterface a few lines up, meaning we lose the state tracking?

Reply:
This is done, but together with the fix for the other nit, a significant issue was uncovered in `data_loss_protect` handling.

It looks like the `Node` `Drop` implementation does a serialization round-trip, which ends up trying to broadcast a revoked commitment tx, since we restored an old state.  This seems to be an actual issue with channel reestablish - not sure how you intended it to recover going forward since it's out of sync.  It seems like if there's data-loss, we have to update our commitment number instead of just erroring in `channel_reestablish`?

Reply:
Hmm, can you push the branch that demonstrates that somewhere? I suppose the "right" answer could be to not panic, return an Err, make sure `ChannelMonitor` handles that fine (it should?) and then hope we can broadcast the current state when our counterparty broadcasts theirs. That's not quite enough because we should be able to broadcast the old state manually, but then it'd be up to the signer to accept a stale state if we give up waiting on our counterparty.

Reply:
I suppose we could tweak the signer to check a boolean whether we should panic or not and in a few tests (like this one), let it Err instead of panicing and then test that we can still claim a counterparty-broadcasted tx or retry broadcasting our state later.

Reply:
Sorry, pushed now, forgot yesterday.

Reply:
It seems like full support for `data_loss_protect` would entail putting the system in a state where it doesn't try to go on-chain, because we don't have the needed data for a non-revoked tx.  Punting on that for later, I've added a commit that marks the node as failed and doesn't try to execute `Node::Drop` for the node that suffered the data-loss in `test_data_loss_protect`.

Reply:
As discussed elsewhere, we now return `Err` when there is a revocation issue and `err_on_sign_revoked_holder_tx` is on in `TestKeysInterface` / `EnforcingChannelKeys`.

Reply:
... however, this workaround no longer works as a result of `OnChainTx.holder_commitment` no longer being optional.  As part of that change, we panic if an `Err` is return when trying to sign the latest commitment.  Also, this is now an issue also in tests that test justice reactions by simulating a bad actor - they panic during HTLC claims when they react to their own broadcast.

I see a couple of approaches:

- propagate the `Err` up the call stack and don't panic
- simulate bad actor / data loss more accurately by rolling back the signer or at least the revocation counter

The rebase that fails tests was pushed just now.

Reply:
Actually, the simplest fix is to disable the revocation enforcement for these spots.